hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -204,7 +252,9 @@ public int getFactoryId() {
 
     @Override
     public int getClassId() {
-        return SqlDataSerializerHook.QUERY_DATA_TYPE;
+        return isCustomType()
```

Comment:
this check does not know about the protocol version, so uses new ids even if `out.getVersion()` is 5.3. That may cause `SqlCatalogCompatibilityTest` failures

Reply:
There are two ways to make `getClassId()` aware of the cluster version:
1. Introduce the following interface and modify `EnterpriseDataSerializableSerializer`.
    ```java
    public interface VersionedIdentifiedDataSerializable extends IdentifiedDataSerializable, Versioned {

        int getClassId(Version clusterVersion);

        default int getClassId() {
            throw new UnsupportedOperationException();
        }
    }
    ```
    ```java
    private void writeIdentifiedDataSerializable(...) { // EnterpriseDataSerializableSerializer
        ...
        out.writeInt(obj.getFactoryId());
        out.writeInt(obj instanceof VersionedIdentifiedDataSerializable
                ? ((VersionedIdentifiedDataSerializable) obj).getClassId(clusterVersion)
                : obj.getClassId());
        ...
    }
    ```
2. Access `HazelcastInstanceFactory.INSTANCE_MAP.values().iterator.next().get().getCluster().getClusterVersion()`.

I think _versioning_ feature should allow changing class ID (and maybe even factory ID) since it is the user's responsibility to provide and maintain the `DataSerializableFactory`. Because `Versioned` interface can be used for `DataSerializable`s as well, we cannot modify it. That's why I propose to introduce `VersionedIdentifiedDataSerializable`. The second solution might be appealing due to its locality, but I consider it a hack. What do you think?

Reply:
Option 1 + not assuming that QDT instances are singleton (due to MAPPINGs created in <=5.3) is the way to go in 5.4.

Reply:
5f47d406 + hazelcast/hazelcast-enterprise@83d076c3